### DISCUSSION TITLE
Верхний ряд

### DISCUSSION THREAD

**Discussion by tipodanet** [k-ivan/visual-bookmarks-chrome#170] [Ideas]
Идея не новая . Такое есть у многих Визуальных. Я так понимаю, что проблемы с созданием Папок нет -- расширение берет все закладки и папки из самого браузера. C Chrome все понятно -- он берет закладки и Аккаунта юзера Google Bookmarks. 
На других браузерах на Хромиуме я не проверял. Даже не представляю, будет там работать Ваше расширение или нет. Не знаю механизма источника. 
Но вот ВЕРХНИЙ РЯД есть у некоторых расширений и у браузеров со своими Speed Dial . И это удобно -- сразу выходить на начальную страницу,  и сразу видеть нужную стопку по нужной теме.

Т.к. расширение не контролирует создание папок, то 
Предложение: сделать **_верхний ряд равным верхним папкам._** .
Ряд без картинок, только названия сайтов. И значит, без ссылок. 
И если в корне всех закладок встречается чистая закладка (без папки) , то ее не учитывать (юзер должен сам ее перенести в существующую папку).
Понятно, чтобы он был постоянным на всех  уровнях вложенности и это можно было ВКЛ/ВЫКЛ.

**Comment by k-ivan** [comment]
Расширение работает на Blink браузерах, Chrome, Edge, Opera, Brave(вроде). И так же на Firefox.

Что значит верхний ряд, что это такое, ничего не понял. Можно более подробно?

**Comment by tipodanet** [comment]
Недавно MS ввели смайлы и картинки везде в тексте (Win +точка). Там много иконок и смайлов. Можно в эти папки в верхнем ряду  вначале сделать иконки перед названием. Плюс цвет background-color.

**Reply by k-ivan** [reply]
О каких папках речь, которые внутри селекта(рядом с поиском), если да, то туда нельзя добавить, это системный селект, его нельзя особо кастомизировать. Но если Вы укажите папке в названии эмоджи, то он скорее всего отрисуется в селекте.

**Comment by tipodanet** [comment]
![image](https://github.com/user-attachments/assets/40db2fd7-22e5-4ba1-b90f-6ccaad3160b1)

**Reply by tipodanet** [reply]
Это верхний уровень  . Просто копия названий папок верхнего уровня. Красный ряд постоянный.
Ширина по навзанию. Если у меня 3 ряда папок, 18 штук, то верхний ряд вполне может вместить 18 названий папок.

**Reply by k-ivan** [reply]
Не понял, они дублируются ещё и сверху?

Вы предлагаете просто дублировать папки сверху?

**Reply by tipodanet** [reply]
Да. именно дублируются. У Вивальди это есть. Скрин нужен ?

**Reply by tipodanet** [reply]
Здесь главное -- постоянство. Если я закопался на 10 уровень вглубь, то сразу выход наверх, в нужную папку. Удобно

**Reply by k-ivan** [reply]
Пока не вижу в этом смысла, ради чего?

**Reply by tipodanet** [reply]
![image](https://github.com/user-attachments/assets/42fc075e-2207-4b86-8311-2e827be2d12c)

**Reply by tipodanet** [reply]
Это есть и у других Визуальных. Так что здесь нет кражи интеллектуальной собственности.

**Reply by k-ivan** [reply]
А, стало понятно по скрину.
Пока есть селект с папками на верхнем уровне.

Да такой вариант был в планах(правда не срочных), там надо хорошенько обдумать, как это реализовать.

**Reply by tipodanet** [reply]
я считаю, что у вас уже все есть , все кнопки выхода наверх уже есть. 
Этот ряд никак не отражается на основном поле. Т.е. ничего переделывать и не надо.
Ряд сам формируется из имен папок в корне закладок.

**Reply by tipodanet** [reply]
Прокрутка вправо/влево , со стрелками. У кого много папок в корне. 

```
padding-left: 5px, 
padding-right: 5px . 
border: 1px solid white; 
border-radius: 2px; 

```
Вот и все настройки этого ряда :) .

**Reply by k-ivan** [reply]
Там много нюансов, как его сделать удобным, скорее всего потребуется избавиться от селекта, много мелочей короче

**Reply by k-ivan** [reply]
> Прокрутка вправо/влево , со стрелками. У кого много папок в корне. 
> 
> ```
> padding-left: 5px, 
> padding-right: 5px . 
> border: 1px solid white; 
> border-radius: 2px; 
> 
> ```
> Вот и все настройки этого ряда :) .

Вы сильно ошибайтесь

**Reply by tipodanet** [reply]
В Вивальди он неотключаемый. 
Я  как главный поборник чистого дизайна могу предположить , что кому-то он может помешать воспринимать красоту квадратов. 
Мне он нужен ради скорости выхода наверх. 
Значит, надо учесть интересы всех. Т.е. галку в настройках. 
Предположу, что там просто галка для видимости 
`visibility: visible (hidden);`
т.к. с hidden место под ряд будет уничтожаться с поднятием всего поля .

**Reply by k-ivan** [reply]
Это штука должна быть опциональной конечно, без всяких визибилити. То есть код с ней вообще не должен грузиться, если она выключена.

Там много нюансов, нужно будет перерабатывать нынешнюю шапку, а возможно разделять поиск и меню с папками на отдельные сущности. Это не просто паддингт и радиус задать.

**Comment by tipodanet** [comment]
Я понял , какой Селект. У вас он тоже Expand :) ))) Вы не знаете , сколько ЛЕТ юзеры на форуме вивальди просят  "Collaps All Folder !!"
. Там даже в подписи у некоторых это написано . Разрабам пофиг, они на свой форум не ходят. 
Представьте, как выглядит селект с 10 000 закладок. 
Этой проблеме лет 6 , или больше. И ответ представителей, кто по-русски говорит у юзерами (похоже, которых там тоже никто не слушает, посылают их)  -- У нас есть более важные задачи. 
Иван, вы можете сами оценить сложность решения этого вопроса. Сколько вам  надо минут на сворачивание всех папок селекта ? 6 минут достаточно ?

**Reply by k-ivan** [reply]
Селект конечно будет быстрее, тем более нативный(системный). А вот огромное кол-во папок отрендеренных в меню, будут тормозить и жрать памяти в разы больше. И удобство будет сомнительное, когда их будет очень много. Да, нативный селект не такой красивый, но производительность и в целом удобство будет за ним, без шансов.

Он уступит только в визуальном виде.

**Reply by k-ivan** [reply]
Чтобы свернуть селект и развернуть требуется 1 секунда

**Reply by tipodanet** [reply]
![image](https://github.com/user-attachments/assets/f496fc43-590e-4ed0-873f-5f4a9cef7077)
Справа белая точка -- это scrollbar для 5000 закладок.
вы видите там кнопку "Collapse / Expand" ?  Нет. Вот ее юзеры и просят 6 лет, или больше :) 
Вот как выглядит ориентированность на своего любимого пользователя.

**Reply by k-ivan** [reply]
Ну это справедливости ради и не селект. Я говорил про селект в своем расширении.

**Reply by tipodanet** [reply]
я не знаю точно, но где-то есть свойство для папок , может не CSS , именно так и называется "Collapse / Expand" . Я его где-то видел. Т.е. все уже есть. 
И  у Вашего Селекта  тоже донжны быть такие кнопки. Станлартные Плюс/Минус  рядлом с папкой это решают, или треугольник Вниз/Вправо. Решений много, даже не надо кнопки рисовать.

**Reply by k-ivan** [reply]
Есть summary > details тэги. Но я не понимаю причем мое расширение и проблема в Вивальди.

Повторюсь, системный селект не кастомизируется, не получится вставить туда логику или визуал особенный.

**Reply by tipodanet** [reply]
При том, что Вам тоже надо свернуть ваш Селект. Он должен сворачиваться к корневым папкам. И запоминать состояние. Есть вероятность что юзер сидит в одной папке, и редко выходит из нее. Ему нужно сохранения состояния. 
В данном виде Селект малопригоден для большого числа Закладок.

**Reply by k-ivan** [reply]
Ваш вариант ниразу не удобнее. Более того,
чтобы сделать древовидную структуру, которая сможет скрываться, раскрываться и ещё учесть, что папок будет много, он должен быть динамическим и не рендерить, то что не раскрыто, а это тоже нетривиально. А если рендерить все и сразу можно столкнуться с тормозами в интерфейсе.

**Reply by tipodanet** [reply]
Я только сейчас узнал, что есть  "системный селект" и не системный. Я бы предпочел тот  Селект, который сворачивается и запоминает состояние. Его внешний вид вторичен.

**Reply by k-ivan** [reply]
Для меня важнее производительность и отзывчивость интерфейса, чего не даст кастомный селект, но даст системный

**Reply by tipodanet** [reply]
Хорошо, представим не 5000 закладок, а хотя бы 500. Системный Селект ничего не рендерит, и показывает все 500 сразу. Зато быстро. тогда просьба добавить радио-кнопку в Настройках для вкл/выкл Селекта.

**Reply by k-ivan** [reply]
Не обещаю, но подумаю над этим, чтобы его выключать придется скорее всего так же разделять его с поиском. Так сейчас это единый компонент.

**Reply by k-ivan** [reply]
Возможно (не факт) и меню из папок добавлю, но точно не в приоритете.

Вот сейчас в обсуждении я уже начал сомневаться в его нужде, так как большое кол-во папок + закладок в текущей папке реально приведет к жору ресурсов и возможным тормозам.

**Reply by tipodanet** [reply]
Какая связь между 10 корневыми  папками и тем, что внутри 5000 или 500 закладок ? 
Верхний ряд не знает о том, сколько всего закладок. Его интересует только 10 корневых. 
Корневые папки и сейчас  видны вместе с корневыми закладками . Только в виде квадратов. То же количество. 10 штук. И Сейчас все расширение не знает о том, сколько всего закладок. 
Если сейчас у вас нет ограничения на число корневых папок, то откуда могут появиться тормоза от наличия верхнего ряда ?

**Reply by k-ivan** [reply]
Есть ограничение, я беру сразу топ левел папки, в разных браузерах ситуация может разниться, но в сумме их будет 3-5 всего.

**Reply by k-ivan** [reply]
И да, верхний уровень папок, это не то что Вы видите в папке открытой по умолчанию, то что например хранится в папке "панели закладок" - это не верхний уровень. В селекте я строю древовидный вид, чтобы отобразить всех папки в приложении(не только верхний уровень, так там всего 3 папки). В горизонтальном меню это уже не прокатит, и там нужно думать как быть.

Что еще говорит, о большом кол-ве подводных камней и всяких нюансах. Я думаю дальше спорить смысла нет.

**Reply by k-ivan** [reply]
Если я придумаю, как лучше это реализовать, то добавлю. Пока все варианты кажутся очень затратными и неудобными, после селекта.

**Reply by tipodanet** [reply]
Верхний ряд — это не меню, там не предполагается выпадающего списка. Это просто ряд корневых папок  без закладок. Просто клик, открывает содержимое кликнутой корневой папки

**Reply by k-ivan** [reply]
Я писал чуть выше, что корневых папок(например в хроме) всего 3.

Остальные папки, создаются уже внутри корневых, и корневыми не являются.

**Reply by tipodanet** [reply]
А понял, да, там есть "другие закладки", "мобильные закладки" и "панель закладок". Можно их назвать корневыми. 
Но приложение не показывает первые 2. Сейчас все как надо. Сейчас корневые , которые видит юзер, это только "панель закладок". 
Вся тема о том, чтобы повторить эти папки сверху без картинки. Для быстрого выхода на "рабочий стол"

**Reply by k-ivan** [reply]
Приложение как раз показывает все 3 вида
